### ISSUE TITLE
Symbolic model of the FF dispatch layer (big bite — load-bearing for the weights-are-the-polynomial claim)

### ISSUE THREAD

**Issue by oaustegard** [oaustegard/llm-as-computer#69]
# Symbolic model of the FF dispatch layer (load-bearing for the "weights ARE the polynomial" claim)

Follow-up to #65. The merged symbolic executor (PR #66) and catalog runner
(PR #67) prove one claim: *the Python-level opcode semantics compose into a
polynomial over pushed constants*. That's genuine but limited — it's a claim
about the ISA, not about the transformer weights. This issue closes the gap.

## What the current story actually proves

Read `executor.py:503-876` (`CompiledModel._compile_weights` and
`CompiledModel.forward`) with `symbolic_executor.py` open alongside. The
compiled transformer has two dispatch paths:

- **Linear path** (`executor.py:670-703, 800-811`). `M_top` routes one of
  `[arg, val_a, val_b, val_c, local_val, heap_val]` to the output per opcode.
  For `PUSH, POP, DUP, HALT, JZ, JNZ, NOP, SWAP, OVER, ROT` this is a single
  1.0 entry — a pure linear map on inputs that were already extracted by the
  attention heads. No arithmetic; just routing.
- **Nonlinear path** (`executor.py:813-870`). For `ADD, SUB, MUL, DIV_S,
  REM_S, EQ, LT_S, GT_S, AND, OR, XOR, SHL, CLZ, CTZ, POPCNT, ABS, NEG, ...`
  the FF dispatch **falls through to Python**: `nonlinear[ADD] = float((va +
  vb) & MASK32)`, etc. `M_top` rows for these opcodes are zero.

So the compiled transformer's `M_top` can faithfully route arithmetic
**results**, but the arithmetic itself is computed by CPython int ops inside
the forward pass, not by matmul. The thesis slogan "weights are a compiler
target; the forward pass is a CPU" is currently supported by the linear
path only — for `ADD, SUB, MUL` the CPU is literally the host.

The symbolic executor inherits this gap. `Poly.__add__`, `Poly.__mul__`, etc.
are Python polynomial algebra. The symbolic executor's claim is "this
polynomial is what the program computes" — true at the ISA level, silent
about the weights.

## What this issue does

Replace the Python-arithmetic nonlinear path with a bilinear / polynomial
form over embeddings, for at least `ADD`, `SUB`, `MUL`. Prove algebraically
and verify numerically that:

1. There exist analytically-set weight matrices `M_ADD`, `M_SUB`, `M_MUL`
   (or a shared bilinear-form machinery) such that for all integers `a, b`
   in a specified range, the FF output on `(E(a), E(b))` equals `E(a op b)`.
2. `E` is the same embedding used by the attention heads for `DIM_VALUE`.
3. When fed the program trace, the modified `CompiledModel` produces the
   same per-step top-of-stack as the current model on the catalog's
   collapsed programs.
4. The symbolic executor's output polynomial is literally what the bilinear
   form computes symbolically — `forward_symbolic(PROG, (x0, x1, ...))` and
   `SymbolicExecutor(PROG).poly.eval_at(x0, x1, ...)` produce the same
   expression, not just the same number.

Point 4 is the load-bearing claim: **the weight matrix is a lossy encoding
of the formula, not a runtime for it.** Everything else is plumbing.

### In scope

- **Embedding scheme for values.** Keep the current scalar-in-`DIM_VALUE`
  embedding for inputs; introduce an *extended* value embedding `E_mul`
  used only for the bilinear MUL form. `E_mul(a) = (1, a, a^2, ..., a^K)`
  for some small K — a standard polynomial-embedding trick. Document the
  tradeoff: K controls the max degree of polynomials the FF layer can
  realize without composition.
- **`M_ADD`, `M_SUB` as linear maps.** `E(a+b) = E(a) + E(b)` under scalar
  embedding; `M_ADD @ [va, vb] = va + vb` is the existing routing already
  — just move it from the Python nonlinear path to a new linear slot in
  `M_top`. Same for `SUB`. This is the trivial first step and should
  produce zero behavior change; commit it as the first PR so the
  harder MUL work lands on a clean baseline.
- **`M_MUL` as a bilinear form.** Introduce `B_MUL: R^d × R^d → R` such
  that `B_MUL(E(a), E(b)) = a·b` for integers `a, b` in
  `[-2^15, 2^15-1]` (or a larger range — pick one and pin it). Document
  the construction: for scalar embedding it's literally `B_MUL = e_val ⊗
  e_val` so `B_MUL(x, y) = x·y` directly. The "bilinear form" language
  matters because it's what a standard FF layer with a gated/bilinear
  variant can actually compute — this is the bridge to the "FF layer is a
  polynomial evaluator" claim.
- **Symbolic forward pass.** A new `CompiledModel.forward_symbolic(...)`
  that takes symbolic `Poly` inputs where the current code takes
  `torch.tensor(float)`. It applies `M_ADD, M_SUB, M_MUL` as symbolic
  operations on `Poly` objects and returns a `Poly` at `HALT`.
- **Equivalence proof artifact.** A `dev/ff_symbolic_equivalence.md`
  document + a `test_ff_symbolic.py` test file that together show: for
  every catalog program that collapses, `forward_symbolic` and
  `SymbolicExecutor` produce the same `Poly`. Not "numerically match on
  sample inputs" — *structurally* equal after canonical simplification.
- **Honest treatment of i32 wrap.** The current Python path has `& MASK32`
  (WASM i32 overflow). A polynomial over Z does not wrap. Two options,
  pick one and document:
  - **(a) Drop the mask** — produce a polynomial over Z, add a
    `range_check(max_abs_value)` that asserts no wrap would have occurred
    on the catalog inputs. The claim becomes "over Z and within range, the
    weights realize this polynomial".
  - **(b) Polynomials over Z/2^32** — carry the mod-2^32 constraint
    through the algebra. Heavier; probably out of scope, but call it out
    as a follow-up.
  Option (a) is the recommended choice — it matches the spirit of the
  symbolic executor (already pure-integer, no wrap modeling) and keeps
  this issue bounded.

### Non-goals

- **`DIV_S`, `REM_S`, `LT_S`, `AND`, etc.** These require algebras beyond
  the polynomial ring (rationals, piecewise / Heaviside, mod-2). Each is
  its own issue — see the roadmap issue for the list. This issue is done
  when ADD/SUB/MUL are bilinear and the equivalence proof holds.
- **Attention heads.** The linear routing from heads to FF is unchanged;
  this issue only touches the FF dispatch.
- **Training.** No learned weights. All matrices analytically set.
- **Retiring the Python nonlinear path.** Keep it as the reference
  implementation; add the bilinear path alongside, and verify they agree.

### New module: `ff_symbolic.py`

```python
# Minimum surface:
def E(v: int) -> torch.Tensor: ...              # scalar embedding -> d_model vector
def E_inv(emb: torch.Tensor) -> int: ...        # read DIM_VALUE, round

def M_ADD() -> torch.Tensor: ...                # [1, 1] row, stored as d_model x d_model slice
def M_SUB() -> torch.Tensor: ...                # [1, -1]
def B_MUL() -> torch.Tensor: ...                # bilinear form; shape (d_model, d_model)

def forward_add(ea: torch.Tensor, eb: torch.Tensor) -> torch.Tensor:
    """Returns E(a+b) given E(a), E(b). No Python arithmetic."""

def forward_mul(ea: torch.Tensor, eb: torch.Tensor) -> torch.Tensor:
    """Returns E(a*b) via B_MUL(ea, eb). No Python arithmetic."""

# Symbolic versions operate on Poly instead of Tensor:
def symbolic_add(pa: Poly, pb: Poly) -> Poly: ...
def symbolic_mul(pa: Poly, pb: Poly) -> Poly: ...
```

The symbolic and numeric forms **share the same formula** — the
symbolic version substitutes `Poly.__add__` / `Poly.__mul__` into the same
tensor-free expression. That's the equivalence: one spec, two interpreters
(tensors of floats vs. polynomials).

### `CompiledModel.forward_symbolic`

Mirrors `forward()` line-for-line with:
- `val_a, val_b, val_c` as `Poly` instead of `torch.tensor(float)`.
- Attention-head extraction replaced by `stack[SP], stack[SP-1], stack[SP-2]`
  (polynomials on the symbolic stack).
- Linear dispatch unchanged (it's already polynomial-compatible).
- Nonlinear dispatch replaced with `symbolic_add / symbolic_sub /
  symbolic_mul`; other nonlinear ops raise `BlockedOpcode` for now.

### Equivalence theorem (stated; proof is a test + a short writeup)

For every catalog program `P` that currently reports status `collapsed`:

```
SymbolicExecutor(P).poly == CompiledModel().forward_symbolic(P).poly
```

where `==` is structural equality on canonical `Poly`. This is a one-line
test per program (parametrize over the 15 collapsed rows) and the writeup
explains why it holds: both interpreters apply the same formula for ADD /
SUB / MUL (polynomial algebra) to inputs from the same source (stack
reads). The only nontrivial step is establishing that `B_MUL` realizes
polynomial multiplication — which for scalar `DIM_VALUE` embedding is
direct.

### Correctness

- [ ] `ff_symbolic.py` lands with `M_ADD`, `M_SUB`, `B_MUL` and both
      numeric (`forward_add/sub/mul`) and symbolic (`symbolic_add/sub/mul`)
      variants.
- [ ] `CompiledModel.forward` is updated: `ADD`, `SUB`, `MUL` move from
      the Python nonlinear path to use `forward_add/sub/mul`. Other
      nonlinear ops unchanged.
- [ ] `test_consolidated.py` still green — `CompiledModel` produces the
      same trace as `NumPyExecutor` on every program it handles today.
- [ ] `CompiledModel.forward_symbolic` lands, consuming `Poly` inputs and
      producing a `Poly` output at `HALT`.
- [ ] `test_ff_symbolic.py` parametrizes over all 15 currently-collapsed
      catalog programs and asserts structural `Poly` equality between
      `SymbolicExecutor` and `forward_symbolic`.

### Documentation

- [ ] `dev/ff_symbolic_equivalence.md` explains the claim, the embedding
      choice, the i32-wrap decision, and walks through one example
      (`dup_add_chain_x4`: 9 heads → one bilinear form → `16·x0`).
- [ ] `dev/symbolic_collapse_report.md` preamble gains a sentence: "The
      polynomial column is the *structural* output of both the symbolic
      executor and the FF bilinear form on `forward_symbolic` — not just a
      numerical agreement."
- [ ] README blog-posts section links to the forthcoming follow-up blog
      post (TBD title, draft after the PR merges).

### Non-regressions

- [ ] Numeric behavior of `CompiledModel` unchanged for all programs
      (including non-polynomial ones whose nonlinear path stays Python).
- [ ] Parameter count reported correctly — `M_ADD`, `M_SUB`, `B_MUL`
      parameters are counted into the compiled total. The blog post's
      "964 compiled parameters" number will update; that's fine, note the
      delta in the PR description.
- [ ] Symbolic executor's existing tests all pass unchanged.

### Honest limits

- [ ] The equivalence theorem is stated over Z, with a documented range
      assumption that no i32 wrap occurs on catalog inputs (verified at
      test time via `range_check`).
- [ ] The PR description and the equivalence writeup explicitly list what
      this issue does *not* prove: comparisons, division, bitwise, loops,
      branches — all unchanged.

## Why this is the right next big bite

Without this, the LAC repo's claim is narrowly "we compiled the ISA
semantics into a PyTorch module that happens to wrap Python arithmetic".
With this, the claim becomes "for the arithmetic fragment, the weights
*are* the polynomial: the FF layer is a bilinear form, the symbolic
executor is the same bilinear form over `Poly` inputs, and the two agree
structurally." That's the sentence the bridging blog post wanted to be
able to write but couldn't.

It also unlocks the honest comparison with eml-sr on its own terms: the
eml-sr tree isn't just "a tree that computes the same function as the
transformer" — it's "a tree that computes the same polynomial that the
transformer's bilinear FF layer realizes directly, via a different
grammar." Three grammars, one object, provably.

## References

- Parent issue: #65
- Symbolic executor: PR #66 (merged)
- Catalog + eml bridge: PR #67 (merged)
- Bridging blog post:
  https://muninn.austegard.com/blog/where-the-computer-meets-the-calculator.html
- Relevant code:
  - `executor.py:503-876` — `CompiledModel._compile_weights`, `forward`
  - `executor.py:670-704` — `M_top` linear routing matrix
  - `executor.py:813-870` — the nonlinear path to replace
  - `symbolic_executor.py` — `Poly`, `SymbolicExecutor`
  - `symbolic_programs_catalog.py` — `classify_program`, `poly_to_expr`

## Out-of-scope follow-ups this enables

Once `ADD/SUB/MUL` have bilinear forms, these become natural extensions
(each a separate issue):

- **Rational arithmetic** for `DIV_S`, `REM_S` (the FF layer gains a
  "rational bilinear form"; Polys become rationals).
- **Piecewise bilinear forms** for comparisons (`LT_S`, `GT_S`, `EQ`) —
  gates into one of two output branches based on a sign indicator.
- **Mod-2 bilinear forms** for bitwise ops (`AND`, `OR`, `XOR`) — different
  algebra, same FF-layer machinery.

None of those are in this issue. They are called out only so the roadmap
is visible.
